### ISSUE TITLE
Feature Request: Add two-way ssl support (client certificates)

### ISSUE THREAD

**Issue by stevenpack** [cefsharp/CefSharp#671]
There's currently nothing available in the API to add client certficates to requests. This prevents CefSharp being used in any secure environment where client certificates are used.

Something like 

    public X509Certificate2Collection GetCertificates(IWebBrowser browser, IRequest request)
    {
         //return default collection, or popup OS prompt to select...
    }

**Comment by jimbourekas** [comment]
So,does this mean that it's a really hard to-do?

**Comment by jornh** [comment]
You'll likely have better luck asking what's required in the CEF project. Exposing the CEF API for it in CefSharp is probably easy.

**Comment by amaitland** [comment]
If .NET supports client certifices you could implement a schema handler and proxy the requests using web client to load the data and cef purely for display. It's an untested theory.

**Comment by jimbourekas** [comment]
Ok, I starred the upstream CEF issue mentioned in https://github.com/cefsharp/CefSharp/issues/671 as per Jorns pointer (https://groups.google.com/forum/#!topic/cefsharp/0WsTHpB1ZXs)

Cheers,

jim

**Comment by amaitland** [comment]
https://bitbucket.org/chromiumembedded/cef/issue/1608/add-support-for-client-certificate-ssl

**Comment by amaitland** [comment]
Support has been added to `CEF` `master` branch, so likely be in version `45` (no schedule for that yet). There is of course always the option of building `CEF` from `source`.

Relevant commit.
https://bitbucket.org/chromiumembedded/cef/commits/0d905e3

**Comment by amaitland** [comment]
Looks like this was merged into `2357` (`43`)
https://bitbucket.org/chromiumembedded/cef/issue/1608/add-support-for-client-certificate-ssl#comment-18239112

**Comment by Paulo-Faustino** [comment]
Hey guys, is the client certificate issue #1608 already merged into CEFSharp? I just gave it a shot here but the page that requests a certificate didn't load.

Looking forward,
Thanks

**Comment by jankurianski** [comment]
@Paulo-Faustino No.  First CefSharp must be upgraded to Chromium `43`.  There is a lot of good work being done by @amaitland and @rassilon to help move to `43`.  You can keep track of their progress here: https://github.com/cefsharp/cef-binary/issues/28

Once we move to `43`, would you be interested in working on a PR to expose client certificate support in CefSharp?

**Comment by amaitland** [comment]
> Once we move to 43, would you be interested in working on a PR to expose client certificate support in CefSharp?

> The client certificate must be pre-installed using OS functions as described above. The selection is then completely automatic (it will use the first client certificate that matches).

It shouldn't actually require any changes, just an upgrade to a supporting `CEF` version. Still waiting on `cefbuilds.com` to catch up.

**Comment by jankurianski** [comment]
> It shouldn't actually require any changes, just an upgrade to a supporting CEF version. Still waiting on cefbuilds.com to catch up.

Ahh, I see.  Thanks for clarifying @amaitland :+1:

**Comment by Paulo-Faustino** [comment]
So you are saying that as soon as they catchup on the build at cefbuilds.com we'll have cefsharp updated?
Thanks!

**Comment by amaitland** [comment]
> So you are saying that as soon as they catchup on the build at cefbuilds.com we'll have cefsharp updated?

Not quite. When there is the latest build on `cefbuilds.com` I'll look at publishing a new version of the `cef.redist` project, which we'll then upgrade the current `2272` branch (also renaming it to `2357` in the process). Hopefully next week sometime. That won't change the fact that there is no official release yet, that's likely a few months away.

**Comment by Paulo-Faustino** [comment]
@jankurianski I'm interested.
@amaitland Looking forward

**Comment by amaitland** [comment]
`master` has been updated to `2357`, if anyone has time and your interested in checking this out, would be great if someone can report back.

**Comment by Paulo-Faustino** [comment]
Is the NuGet package (CefSharp.Common, CefSharp.WInforms) up-to-date or do I have to download and build the sources. What about the redist, where should I download them from in order to test this, NuGet?
Thanks

**Comment by jornh** [comment]
@Paulo-Faustino should be in the continuous integration feeds on https://www.myget.org/gallery/cefsharp

**Comment by amaitland** [comment]
As the changes have been made `upstream` and there is no `API` update required for `CefSharp`, I'm closing this.

If anyone encounters a problem then probably best to raise the issue `upstream` (of course keep us in the loop).
